Add support for including logs from reverted transactions

reductionista opened the pull request:
### Description

JIRA: [NONEVM-1387](https://smartcontract-it.atlassian.net/browse/NONEVM-1387)

Logs from reverted transactions will now be included in LogPoller's logs table if `include_reverted` is set on the Filter object passed to `RegisterFilter`.

### Resolves Dependencies
- https://github.com/smartcontractkit/chainlink/pull/16731

[NONEVM-1387]: https://smartcontract-it.atlassian.net/browse/NONEVM-1387?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

reductionista committed:
Add logpoller.Log.Error, and logpoller.Filter.IncludeReverted

reductionista committed:
Fix orm tests broken by https://github.com/smartcontractkit/chainlink-solana/commit/9ece21b824a824c7e9105b1dadd5f34bd83f877e

reductionista committed:
Add tests for processing reverted logs

dhaidashenko commented on pkg/solana/logpoller/models.go:
```diff
@@ -5,18 +5,19 @@ import (
+	MaxLogsKept     int64
+	IsDeleted       bool // only for internal usage. Values set externally are ignored.
+	IsBackfilled    bool // only for internal usage. Values set externally are ignored.
+	IncludeReverted bool
```
We should also take into account `IncludeReverted` in `MatchSameLogs` method to enforce explicit filter removal if this flag is changed.

reductionista replied:
Could you elaborate on what you mean by this?
> Should we have some protection from cases when a caller registers two equal filters, but with different names and one of them allows reverted transactions?

What's wrong with having two different filters where one matches reverted transactions and one doesn't? Isn't that a valid use case?

> We should also take into account IncludeReverted in MatchSameLogs method to enforce explicit filter removal if this flag is changed.

I think taking into account IncludeReverted in MatchSameLogs would have the opposite effect. Presently, changing the flag forces explicit filter removal (because they are treated as matching, since the flag is ignored), and properly updates the filter with the new value of IncludeReverted. (This is done in one of the tests.). If we changed it to take into account IncludeReverted then it would error out with ErrFilterNameConflict. 

I think the current behavior is what we want.

dhaidashenko reviewed (commented):
Should we have some protection from cases when a caller registers two equal filters, but with different names and one of them allows reverted transactions?

EasterTheBunny commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -146,7 +147,24 @@ func (lp *Service) Process(ctx context.Context, programEvent ProgramEvent) (err
+
 	for filter := range matchingFilters {
+		var revertErr error
+		if blockData.Error != nil {
```
This block looks like all errors are treated as revert errors. Is there not a special RPC error that we need that captures the `reverted` state?

reductionista replied:
Oh, that's a good point. I was under the impression that the terms "failed transaction" and "reverted transaction" were interchangeable. But apparently, that's only true for evm... in Solana it looks like things such as running out of gas are treated as failed transactions but not reverted transactions.

I guess I should rename reverted to failed everywhere?

reductionista replied:
Actually that's not quite right.  On evm, reverted transactions are transactions that fail while executing on chain. These are the only type of failed transactions which get included in blocks (although the logs themselves are just discarded on evm). We don't care about transactions that fail before they're even executed.

So "logs from reverted transaction" is an accurate description of what we want in evm terminology.  On Solana, there is a similar difference between transactions that fail before execution and ones that fail on chain. Apparently, there is no such thing at all as a reverted transaction in Solana, but... there doesn't appear to be a suitable substitute term for transactions that fail while executing.  Suggestions from Gemini AI were "runtime failure" or "execution failure" but these seem too verbose. I think anyone familiar with evm terminology should know what is meant here, so I'd lean towards not renaming it.

prashantkumar1982 commented on pkg/solana/logpoller/log_poller.go:
```diff
@@ -146,7 +147,24 @@ func (lp *Service) Process(ctx context.Context, programEvent ProgramEvent) (err
+				revertErr = fmt.Errorf("unknown type: %v", e)
+				lp.lggr.Warnf("received unknown revert error type %T for log with blockData=%v", revertErr, blockData)
+			}
+			if !filter.IncludeReverted {
```
nit: Probably better if this check is done at the start of the if block for Error.

if (blockData.Error != nil && filter.IncludeReverted) {

reductionista replied:
Yeah, I had it that way originally but I was thinking this might help us gather some information about what sorts of types and errors can be returned here (without having to wait for a Chainlink transaction to error out) by logging this message for each new error type we receive:
```
lp.lggr.Warnf("received unknown revert error type %T for log with blockData=%v", revertErr, blockData)
```
I spent a while looking through the solana-go code to figure out what the possiilities are here, but the type is `interface{}` and the comment next to it links to a 404 Not Found file within an archived Rust repository. I figured since we're not using this information yet for anything, at this stage we can just be saving and logging it:
https://github.com/gagliardetto/solana-go/blob/main/rpc/types.go#L185-L187

I guess I could try and run some analysis of actual on-chain data to see what the possibilities are, but I'm not confident we could categorize all of them easily.

reductionista replied:
I pulled some sample blocks from an rpc server and looked over the errors in failed transactions. It appears that they don't look anything like what I was assuming, so I've replaced the error type with *string (so it can be saved as NULL in the error column of the db for success or include a string representation of the error for failure.

The errors I'm seeing all look something like this:
```
      "err": {
        "InstructionError": [
          2,
          {
            "Custom": 6003
          }
        ]
      }
```
Not particularly useful I guess, but we may as well save it. It looks like what happens is it just gets json unmarshalled into the interface{} type, which creates a map with "InstructionError" as the only key. I figure just reversing the marshalling and storing it as a json string makes the most sense.

Also, took your suggestion and moved the continue up to the top of the block; now that I'm more confident about what type of errors there are, I don't think we need it.

prashantkumar1982 reviewed (commented):
Sorry, clicked on approve by mistake

prashantkumar1982 reviewed (dismissed):
Removing approval

reductionista committed:
Update parsing of revert error

reductionista committed:
Address RegisterFilter PR commments

- Disallow more than one value of IncludeReverted to be registered at once for a
  given (chain_id, address, event_sig) combination. If these values are
  inconsistent then some plugins may be unaware that reverted logs might
  be returned from QueryKey

- Trigger a backfill when updating IsReverted from false to true

reductionista committed:
Fix lint

dhaidashenko commented on pkg/solana/logpoller/filters.go:
```diff
@@ -119,9 +119,26 @@ func (fl *filters) RegisterFilter(ctx context.Context, filter Filter) error {
 
 		fl.removeFilterFromIndexes(*existingFilter)
+	} else {
+		// ensure that the value of IncludeReverted isn't different from any other filters with the same address and event type
```
We'll end up with two similar filters and different IncludeReverted, if there were two filters and I've changed IncludeReverted for one of them.

reductionista committed:
Handle potential conflict while updating IncludeReverted
